File: Source/Engine/Graphics/D3D12/D3D12Texture.cpp
// Copyright © Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

#include "D3D12Texture.h"
#include "D3D12Buffer.h"
#include "D3D12CommandBuffer.h"
#include "D3D12Graphics.h"
#include "directx/d3dx12.h"

namespace Alimer
{
    namespace
    {
        [[nodiscard]] constexpr D3D12_RESOURCE_DIMENSION ToD3D12(TextureType type)
        {
            switch (type)
            {
                case TextureType::Texture1D:
                    return D3D12_RESOURCE_DIMENSION_TEXTURE1D;

                case TextureType::Texture2D:
                case TextureType::TextureCube:
                    return D3D12_RESOURCE_DIMENSION_TEXTURE2D;

                case TextureType::Texture3D:
                    return D3D12_RESOURCE_DIMENSION_TEXTURE3D;
                default:
                    ALIMER_UNREACHABLE();
                    return D3D12_RESOURCE_DIMENSION_UNKNOWN;
            }
        }

        [[nodiscard]] constexpr DXGI_FORMAT GetTypelessFormatFromDepthFormat(PixelFormat format)
        {
            switch (format)
            {
                case PixelFormat::Depth16Unorm:
                    return DXGI_FORMAT_R16_TYPELESS;
                case PixelFormat::Depth32Float:
                    return DXGI_FORMAT_R32_TYPELESS;
                case PixelFormat::Depth24UnormStencil8:
                    return DXGI_FORMAT_R24G8_TYPELESS;
                case PixelFormat::Depth32FloatStencil8:
                    return DXGI_FORMAT_R32G8X24_TYPELESS;

                default:
                    ALIMER_ASSERT(IsDepthFormat(format) == false);
                    return ToDXGIFormat(format);
            }
        }
    }

    D3D12Texture::D3D12Texture(D3D12Graphics& device, const TextureCreateInfo& info, const void* initialData, ID3D12Resource* existingHandle, PixelFormat viewFormat)
        : Texture(info)
        , device{ device }
        , D3D12GpuResource(existingHandle, D3D12_RESOURCE_STATE_COMMON)
    {
        if (existingHandle == nullptr)
        {
            D3D12MA::ALLOCATION_DESC allocationDesc{};
            allocationDesc.HeapType = D3D12_HEAP_TYPE_DEFAULT;

            D3D12_RESOURCE_DESC resourceDesc = {};
            resourceDesc.Dimension = ToD3D12(info.type);
            resourceDesc.Alignment = 0;
            resourceDesc.Width = info.width;
            resourceDesc.Height = info.height;
            resourceDesc.MipLevels = info.mipLevels;
            resourceDesc.Format = ToDXGIFormat(info.format);
            resourceDesc.SampleDesc.Count = D3D12SampleCount(info.sampleCount);
            resourceDesc.SampleDesc.Quality = 0;
            resourceDesc.Layout = D3D12_TEXTURE_LAYOUT_UNKNOWN;
            resourceDesc.Flags = D3D12_RESOURCE_FLAG_NONE;

            if (info.type == TextureType::TextureCube)
            {
                resourceDesc.DepthOrArraySize = info.depthOrArraySize * 6;
            }
            else
            {
                resourceDesc.DepthOrArraySize = info.depthOrArraySize;
            }

            if (CheckBitsAny(usage, TextureUsage::Storage))
            {
                resourceDesc.Flags |= D3D12_RESOURCE_FLAG_ALLOW_UNORDERED_ACCESS;
            }

            D3D12_CLEAR_VALUE clearValue = {};
            D3D12_CLEAR_VALUE* pClearValue = nullptr;

            if (CheckBitsAny(usage, TextureUsage::RenderTarget))
            {
                // Render targets and Depth/Stencil targets are always committed resources
                allocationDesc.Flags = D3D12MA::ALLOCATION_FLAG_COMMITTED;

                clearValue.Format = resourceDesc.Format;

                if (IsDepthStencilFormat(format))
                {
                    state = D3D12_RESOURCE_STATE_DEPTH_WRITE;
                    resourceDesc.Flags |= D3D12_RESOURCE_FLAG_ALLOW_DEPTH_STENCIL;
                    if (!CheckBitsAny(usage, TextureUsage::Sampled))
                    {
                        resourceDesc.Flags |= D3D12_RESOURCE_FLAG_DENY_SHADER_RESOURCE;
                    }

                    clearValue.DepthStencil.Depth = 1.0f;
                }
                else
                {
                    state = D3D12_RESOURCE_STATE_RENDER_TARGET;
                    resourceDesc.Flags |= D3D12_RESOURCE_FLAG_ALLOW_RENDER_TARGET;
                }

                pClearValue = &clearValue;
            }

            // If depth and either ua or sr, set to typeless
            if (IsDepthFormat(format) && CheckBitsAny(usage, (TextureUsage::Sampled | TextureUsage::Storage)))
            {
                resourceDesc.Format = GetTypelessFormatFromDepthFormat(format);
                pClearValue = nullptr;
            }

            HRESULT result = device.GetAllocator()->CreateResource(&allocationDesc,
                &resourceDesc,
                state,
                pClearValue,
                &allocation,
                IID_PPV_ARGS(&handle)
            );

            if (FAILED(result))
            {
                LOGE("Direct3D12: Failed to create texture: {}", std::to_string(result));
                return;
            }

            dxgiFormat = resourceDesc.Format;

            // Upload data.
            if (initialData != nullptr)
            {
                resourceDesc = handle->GetDesc();
                const uint32_t numSubResources = resourceDesc.MipLevels * resourceDesc.DepthOrArraySize;

                std::vector<D3D12_PLACED_SUBRESOURCE_FOOTPRINT> layouts(numSubResources);
                std::vector<UINT64> rowSizesInBytes(numSubResources);
                std::vector<UINT> numRows(numSubResources);
                device.GetHandle()->GetCopyableFootprints(&resourceDesc, 0, numSubResources, 0, layouts.data(), numRows.data(), rowSizesInBytes.data(), &allocatedSize);

                D3D12UploadContext uploadContext = device.ResourceUploadBegin(allocatedSize);
                uint8_t* uploadMemory = reinterpret_cast<uint8_t*>(uploadContext.CPUAddress);

                const uint8_t* srcMem = reinterpret_cast<const uint8_t*>(initialData);
                const uint64_t srcTexelSize = GetFormatBlockSize(info.format);

                for (uint64_t arrayIdx = 0; arrayIdx < resourceDesc.DepthOrArraySize; ++arrayIdx)
                {
                    uint64_t mipWidth = resourceDesc.Width;
                    for (uint64_t mipIdx = 0; mipIdx < resourceDesc.MipLevels; ++mipIdx)
                    {
                        const uint64_t subResourceIdx = mipIdx + (arrayIdx * resourceDesc.MipLevels);

                        const D3D12_PLACED_SUBRESOURCE_FOOTPRINT& subResourceLayout = layouts[subResourceIdx];
                        const uint64_t subResourceHeight = numRows[subResourceIdx];
                        const uint64_t subResourcePitch = subResourceLayout.Footprint.RowPitch;
                        const uint64_t subResourceDepth = subResourceLayout.Footprint.Depth;
                        const uint64_t srcPitch = mipWidth * srcTexelSize;
                        uint8_t* dstSubResourceMem = uploadMemory + subResourceLayout.Offset;

                        for (uint64_t z = 0; z < subResourceDepth; ++z)
                        {
                            for (uint64_t y = 0; y < subResourceHeight; ++y)
                            {
                                memcpy(dstSubResourceMem, srcMem, Min(subResourcePitch, srcPitch));
                                dstSubResourceMem += subResourcePitch;
                                srcMem += srcPitch;
                            }
                        }

                        mipWidth = Max(mipWidth / 2, 1ull);
                    }
                }

                for (uint32_t subResourceIdx = 0; subResourceIdx < numSubResources; ++subResourceIdx)
                {
                    D3D12_TEXTURE_COPY_LOCATION dst = {};
                    dst.pResource = handle;
                    dst.Type = D3D12_TEXTURE_COPY_TYPE_SUBRESOURCE_INDEX;
                    dst.SubresourceIndex = subResourceIdx;

                    D3D12_TEXTURE_COPY_LOCATION src = {};
                    src.pResource = uploadContext.resource;
                    src.Type = D3D12_TEXTURE_COPY_TYPE_PLACED_FOOTPRINT;
                    src.PlacedFootprint = layouts[subResourceIdx];
                    src.PlacedFootprint.Offset += uploadContext.resourceOffset;

                    uploadContext.commandList->CopyTextureRegion(&dst, 0, 0, 0, &src, nullptr);
                }

                device.ResourceUploadEnd(uploadContext);
            }
        }
        else
        {
            handle->AddRef();
            dxgiFormat = handle->GetDesc().Format;
        }

        // Create default view.
        TextureViewCreateInfo viewInfo;
        viewInfo.format = viewFormat;
        const size_t hash = Hash(viewInfo);
        defaultView = CreateView(viewInfo);
        views[hash] = defaultView;

        OnCreated();
    }

    D3D12Texture::~D3D12Texture()
    {
        Destroy();
    }

    void D3D12Texture::Destroy()
    {
        DestroyViews();
        device.DeferDestroy(handle, allocation);
        OnDestroyed();
        D3D12GpuResource::Destroy();
    }

    void D3D12Texture::ApiSetName()
    {
        auto wideName = ToUtf16(name);
        handle->SetName(wideName.c_str());
    }

    TextureView* D3D12Texture::CreateView(const TextureViewCreateInfo& createInfo)
    {
        auto result = new D3D12TextureView(this, createInfo);
        return result;
    }

    D3D12TextureView::D3D12TextureView(_In_ D3D12Texture* texture, const TextureViewCreateInfo& info)
        : TextureView(texture, info)
        , device(texture->GetDevice())
    {
        const TextureUsage usage = texture->GetUsage();
        if (CheckBitsAny(usage, TextureUsage::RenderTarget))
        {
            if (!IsDepthStencilFormat(format))
            {
                rtv = CreateRTV(texture->GetHandle());
            }
            else
            {
                readWriteDSV = CreateDSV(texture->GetHandle(), false, false);
            }
        }

        if (CheckBitsAny(usage, TextureUsage::Sampled))
        {
            srv = device.AllocateDescriptor(D3D12_DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV);

            D3D12_SHADER_RESOURCE_VIEW_DESC srvDesc = { };
            srvDesc.Format = ToDXGIFormat(format);
            srvDesc.Shader4ComponentMapping = D3D12_DEFAULT_SHADER_4_COMPONENT_MAPPING;
            srvDesc.ViewDimension = D3D12_SRV_DIMENSION_TEXTURE2D;
            srvDesc.Texture2D.MostDetailedMip = 0;
            srvDesc.Texture2D.MipLevels = texture->GetMipLevels();
            srvDesc.Texture2D.PlaneSlice = 0;
            srvDesc.Texture2D.ResourceMinLODClamp = 0.0f;
            device.GetHandle()->CreateShaderResourceView(texture->GetHandle(), &srvDesc, srv);

            // Allocate bindless handle
            auto bindlessHandle = device.AllocateSRV();
            device.GetHandle()->CopyDescriptorsSimple(1, bindlessHandle.handle, srv, D3D12_DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV);
            bindless_srv = bindlessHandle.index;
        }
    }

    D3D12TextureView::~D3D12TextureView()
    {
        if (rtv.ptr)
        {
            device.FreeDescriptor(D3D12_DESCRIPTOR_HEAP_TYPE_RTV, rtv);
        }

        if (readWriteDSV.ptr)
        {
            device.FreeDescriptor(D3D12_DESCRIPTOR_HEAP_TYPE_DSV, readWriteDSV);
        }

        if (srv.ptr)
        {
            device.FreeDescriptor(D3D12_DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV, srv);
        }

        // TODO: Free bindless
        if (bindless_srv != kInvalidBindlessIndex)
        {

        }
    }

    const D3D12_CPU_DESCRIPTOR_HANDLE& D3D12TextureView::GetDSV(bool readOnlyDepth, bool readOnlyStencil) const
    {
        if (readOnlyDepth)
        {
            if (readOnlyStencil)
            {
                if (!readOnlyDSV.ptr)
                {
                    readOnlyDSV = CreateDSV(static_cast<D3D12Texture*>(texture)->GetHandle(), true, true);
                }

                return readOnlyDSV;
            }
            else
            {
                if (!readOnlyDepthReadWriteStencilView.ptr)
                {
                    readOnlyDepthReadWriteStencilView = CreateDSV(static_cast<D3D12Texture*>(texture)->GetHandle(), true, false);
                }

                return readOnlyDepthReadWriteStencilView;
            }
        }
        else
        {
            if (readOnlyStencil)
            {
                if (!readWriteDepthReadOnlyView.ptr)
                {
                    readWriteDepthReadOnlyView = CreateDSV(static_cast<D3D12Texture*>(texture)->GetHandle(), false, true);
                }

                return readWriteDepthReadOnlyView;
            }

            return readWriteDSV;
        }
    }

    D3D12_CPU_DESCRIPTOR_HANDLE D3D12TextureView::CreateRTV(ID3D12Resource* resource) const
    {
        uint32_t arrayMultiplier = (texture->GetTextureType() == TextureType::TextureCube) ? 6 : 1;

        D3D12_RENDER_TARGET_VIEW_DESC viewDesc = {};
        viewDesc.Format = ToDXGIFormat(format);

        switch (texture->GetTextureType())
        {
            case TextureType::Texture1D:
                if (texture->GetArraySize() > 1)
                {
                    viewDesc.ViewDimension = D3D12_RTV_DIMENSION_TEXTURE1DARRAY;
                    viewDesc.Texture1DArray.MipSlice = baseMipLevel;
                    viewDesc.Texture1DArray.FirstArraySlice = baseArrayLayer;
                    viewDesc.Texture1DArray.ArraySize = arrayLayerCount;

                }
                else
                {
                    viewDesc.ViewDimension = D3D12_RTV_DIMENSION_TEXTURE1D;
                    viewDesc.Texture1D.MipSlice = baseMipLevel;
                }
                break;
            case TextureType::Texture2D:
            case TextureType::TextureCube:
                if (texture->GetSampleCount() != SampleCount::Count1)
                {
                    if (texture->GetArraySize() > 1)
                    {
                        viewDesc.ViewDimension = D3D12_RTV_DIMENSION_TEXTURE2DMSARRAY;
                        viewDesc.Texture2DMSArray.FirstArraySlice = baseArrayLayer;
                        viewDesc.Texture2DMSArray.ArraySize = arrayLayerCount;
                    }
                    else
                    {
                        viewDesc.ViewDimension = D3D12_RTV_DIMENSION_TEXTURE2DMS;
                    }
                }
                else
                {
                    if (texture->GetArraySize() * arrayMultiplier > 1)
                    {
                        viewDesc.ViewDimension = D3D12_RTV_DIMENSION_TEXTURE2DARRAY;
                        viewDesc.Texture2DArray.MipSlice = baseMipLevel;
                        viewDesc.Texture2DArray.FirstArraySlice = baseArrayLayer * arrayMultiplier;
                        viewDesc.Texture2DArray.ArraySize = arrayLayerCount * arrayMultiplier;
                        viewDesc.Texture2DArray.PlaneSlice = 0;
                    }
                    else
                    {

                        viewDesc.ViewDimension = D3D12_RTV_DIMENSION_TEXTURE2D;
                        viewDesc.Texture2D.MipSlice = baseMipLevel;
                        viewDesc.Texture2D.PlaneSlice = 0;
                    }
                }
                break;
            case TextureType::Texture3D:
                viewDesc.ViewDimension = D3D12_RTV_DIMENSION_TEXTURE3D;
                viewDesc.Texture3D.MipSlice = baseMipLevel;
                viewDesc.Texture3D.FirstWSlice = 0;
                viewDesc.Texture3D.WSize = texture->GetDepth(baseMipLevel);
                break;
            default:
                ALIMER_UNREACHABLE();
                break;
        }

        D3D12_CPU_DESCRIPTOR_HANDLE handle = device.AllocateDescriptor(D3D12_DESCRIPTOR_HEAP_TYPE_RTV);
        device.GetHandle()->CreateRenderTargetView(resource, &viewDesc, handle);
        return handle;
    }

    D3D12_CPU_DESCRIPTOR_HANDLE D3D12TextureView::CreateDSV(ID3D12Resource* resource, bool readOnlyDepth, bool readOnlyStencil) const
    {
        uint32_t arrayMultiplier = (texture->GetTextureType() == TextureType::TextureCube) ? 6 : 1;

        D3D12_DEPTH_STENCIL_VIEW_DESC viewDesc = {};
        viewDesc.Format = ToDXGIFormat(format);

        if (readOnlyDepth)
            viewDesc.Flags = D3D12_DSV_FLAG_READ_ONLY_DEPTH;

        const bool hasStencil = viewDesc.Format == DXGI_FORMAT_D32_FLOAT_S8X24_UINT
            || viewDesc.Format == DXGI_FORMAT_D24_UNORM_S8_UINT;

        if (readOnlyStencil && hasStencil)
            viewDesc.Flags |= D3D12_DSV_FLAG_READ_ONLY_STENCIL;

        switch (texture->GetTextureType())
        {
            case TextureType::Texture1D:
                if (texture->GetArraySize() > 1)
                {
                    viewDesc.ViewDimension = D3D12_DSV_DIMENSION_TEXTURE1DARRAY;
                    viewDesc.Texture1DArray.MipSlice = baseMipLevel;
                    viewDesc.Texture1DArray.FirstArraySlice = baseArrayLayer;
                    viewDesc.Texture1DArray.ArraySize = arrayLayerCount;

                }
                else
                {
                    viewDesc.ViewDimension = D3D12_DSV_DIMENSION_TEXTURE1D;
                    viewDesc.Texture1D.MipSlice = baseMipLevel;
                }
                break;
            case TextureType::Texture2D:
            case TextureType::TextureCube:
                if (texture->GetSampleCount() != SampleCount::Count1)
                {
                    if (texture->GetArraySize() > 1)
                    {
                        viewDesc.ViewDimension = D3D12_DSV_DIMENSION_TEXTURE2DMSARRAY;
                        viewDesc.Texture2DMSArray.FirstArraySlice = baseArrayLayer;
                        viewDesc.Texture2DMSArray.ArraySize = arrayLayerCount;
                    }
                    else
                    {
                        viewDesc.ViewDimension = D3D12_DSV_DIMENSION_TEXTURE2DMS;
                    }
                }
                else
                {
                    if (texture->GetArraySize() * arrayMultiplier > 1)
                    {
                        viewDesc.ViewDimension = D3D12_DSV_DIMENSION_TEXTURE2DARRAY;
                        viewDesc.Texture2DArray.MipSlice = baseMipLevel;
                        viewDesc.Texture2DArray.FirstArraySlice = baseArrayLayer * arrayMultiplier;
                        viewDesc.Texture2DArray.ArraySize = arrayLayerCount * arrayMultiplier;
                    }
                    else
                    {

                        viewDesc.ViewDimension = D3D12_DSV_DIMENSION_TEXTURE2D;
                        viewDesc.Texture2D.MipSlice = baseMipLevel;
                    }
                }
                break;
            case TextureType::Texture3D:
                viewDesc.ViewDimension = D3D12_DSV_DIMENSION_TEXTURE2DARRAY;
                viewDesc.Texture2DArray.MipSlice = baseMipLevel;
                viewDesc.Texture2DArray.FirstArraySlice = 0;
                viewDesc.Texture2DArray.ArraySize = texture->GetDepth(baseMipLevel);
                break;
            default:
                ALIMER_UNREACHABLE();
                break;
        }

        D3D12_CPU_DESCRIPTOR_HANDLE handle = device.AllocateDescriptor(D3D12_DESCRIPTOR_HEAP_TYPE_DSV);
        device.GetHandle()->CreateDepthStencilView(resource, &viewDesc, handle);
        return handle;
    }
}

File: src/Alimer.Graphics/D3D12/D3D12Texture.cs
// Copyright (c) Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

using TerraFX.Interop.DirectX;
using TerraFX.Interop.Windows;
using static Alimer.Graphics.D3D.D3DUtils;
using static TerraFX.Interop.DirectX.DirectX;
using static TerraFX.Interop.DirectX.D3D12_TEXTURE_LAYOUT;
using static TerraFX.Interop.DirectX.D3D12_DSV_DIMENSION;
using static TerraFX.Interop.DirectX.D3D12_HEAP_FLAGS;
using static TerraFX.Interop.DirectX.D3D12_HEAP_TYPE;
using static TerraFX.Interop.DirectX.D3D12_RESOURCE_FLAGS;
using static TerraFX.Interop.DirectX.D3D12_RESOURCE_STATES;
using static TerraFX.Interop.DirectX.D3D12_RTV_DIMENSION;
using static TerraFX.Interop.DirectX.DXGI_FORMAT;
using static TerraFX.Interop.Windows.Windows;
using static Alimer.Utilities.MemoryUtilities;
using DescriptorIndex = System.UInt32;

namespace Alimer.Graphics.D3D12;

internal unsafe class D3D12Texture : Texture, ID3D12GpuResource
{
    private readonly D3D12GraphicsDevice _device;
    private readonly ComPtr<ID3D12Resource> _handle;
    private readonly ComPtr<D3D12MA_Allocation> _allocation;
    private HANDLE _sharedHandle = HANDLE.NULL;
    private readonly D3D12_PLACED_SUBRESOURCE_FOOTPRINT* _footprints;
    private readonly ulong* _rowSizesInBytes;
    private readonly uint* _numRows;
    private readonly void* pMappedData = default;

    private readonly Dictionary<int, DescriptorIndex> _RTVs = new();
    private readonly Dictionary<int, DescriptorIndex> _DSVs = new();

    public D3D12Texture(D3D12GraphicsDevice device, in TextureDescription description, TextureData* initialData)
        : base(description)
    {
        _device = device;
        DxgiFormat = (DXGI_FORMAT)description.Format.ToDxgiFormat();
        bool isDepthStencil = description.Format.IsDepthStencilFormat();

        // If ShaderRead or ShaderWrite and depth format, set to typeless.
        if (isDepthStencil && (description.Usage & TextureUsage.ShaderReadWrite) != 0)
        {
            DxgiFormat = description.Format.GetTypelessFormatFromDepthFormat();
        }

        D3D12_RESOURCE_FLAGS resourceFlags = D3D12_RESOURCE_FLAG_NONE;

        if ((description.Usage & TextureUsage.RenderTarget) != 0)
        {
            if (isDepthStencil)
            {
                resourceFlags |= D3D12_RESOURCE_FLAG_ALLOW_DEPTH_STENCIL;
                if ((description.Usage & TextureUsage.ShaderRead) == 0)
                {
                    resourceFlags |= D3D12_RESOURCE_FLAG_DENY_SHADER_RESOURCE;
                }
            }
            else
            {
                resourceFlags |= D3D12_RESOURCE_FLAG_ALLOW_RENDER_TARGET;
            }
        }

        if ((description.Usage & TextureUsage.ShaderWrite) != 0)
        {
            resourceFlags |= D3D12_RESOURCE_FLAG_ALLOW_UNORDERED_ACCESS;
        }

        D3D12_HEAP_FLAGS heapFlags = D3D12_HEAP_FLAG_NONE;
        bool isShared = false;
        if ((description.Usage & TextureUsage.Shared) != 0)
        {
            heapFlags |= D3D12_HEAP_FLAG_SHARED;
            isShared = true;

            // TODO: What about D3D12_RESOURCE_FLAG_ALLOW_CROSS_ADAPTER and D3D12_HEAP_FLAG_SHARED_CROSS_ADAPTER?
        }

        D3D12_RESOURCE_STATES initialState = description.InitialLayout.ToD3D12();
        if (initialData != null)
        {
            initialState = D3D12_RESOURCE_STATE_COMMON;
        }

        D3D12_RESOURCE_DESC resourceDesc = new();
        resourceDesc.Dimension = description.Dimension.ToD3D12();
        resourceDesc.Alignment = 0;
        resourceDesc.Width = description.Width;
        resourceDesc.Height = description.Height;
        resourceDesc.DepthOrArraySize = (ushort)description.DepthOrArrayLayers;
        resourceDesc.MipLevels = (ushort)description.MipLevelCount;
        resourceDesc.Format = DxgiFormat;
        resourceDesc.SampleDesc = new(description.SampleCount.ToSampleCount(), 0);
        resourceDesc.Layout = D3D12_TEXTURE_LAYOUT_UNKNOWN;
        resourceDesc.Flags = resourceFlags;

        ulong allocatedSize = 0;
        uint numSubResources = Math.Max(1u, resourceDesc.MipLevels) * resourceDesc.DepthOrArraySize;
        _footprints = AllocateArray<D3D12_PLACED_SUBRESOURCE_FOOTPRINT>(numSubResources);
        _rowSizesInBytes = AllocateArray<ulong>(numSubResources);
        _numRows = AllocateArray<uint>(numSubResources);
        device.Handle->GetCopyableFootprints(
            &resourceDesc,
            0,
            numSubResources,
            0,
            _footprints,
            _numRows,
            _rowSizesInBytes,
            &allocatedSize
        );
        AllocatedSize = allocatedSize;

        D3D12_CLEAR_VALUE clearValue = default;
        D3D12_CLEAR_VALUE* pClearValue = null;

        if ((description.Usage & TextureUsage.RenderTarget) != 0)
        {
            clearValue.Format = resourceDesc.Format;
            if (isDepthStencil)
            {
                clearValue.DepthStencil.Depth = 1.0f;
            }
            pClearValue = &clearValue;
        }

        // If shader read/write and depth format, set to typeless
        if (isDepthStencil && (description.Usage & TextureUsage.ShaderReadWrite) != 0)
        {
            pClearValue = null;
        }

        D3D12MA_ALLOCATION_DESC allocationDesc = new();
        allocationDesc.HeapType = D3D12_HEAP_TYPE_DEFAULT;

        HRESULT hr = device.MemoryAllocator->CreateResource(
            &allocationDesc,
            &resourceDesc,
            initialState,
            pClearValue,
            _allocation.GetAddressOf(),
            __uuidof<ID3D12Resource>(),
            _handle.GetVoidAddressOf()
            );

        if (hr.FAILED)
        {
            Log.Error("D3D12: Failed to create Texture.");
            return;
        }

        if (isShared)
        {
            HANDLE sharedHandle = default;
            if (device.Handle->CreateSharedHandle((ID3D12DeviceChild*)_handle.Get(), null, GENERIC_ALL, null, &sharedHandle).FAILED)
            {
                return;
            }

            _sharedHandle = sharedHandle;
        }

        if (!string.IsNullOrEmpty(description.Label))
        {
            OnLabelChanged(description.Label!);
        }

        // Issue data copy on request
        if (initialData != null)
        {
            D3D12_SUBRESOURCE_DATA* data = stackalloc D3D12_SUBRESOURCE_DATA[(int)numSubResources];
            for (uint i = 0; i < numSubResources; ++i)
            {
                ref TextureData subresourceData = ref initialData[i];

                data[i].pData = subresourceData.DataPointer;
                data[i].RowPitch = (nint)subresourceData.RowPitch;
                data[i].SlicePitch = (nint)subresourceData.SlicePitch;
            }

            D3D12UploadContext uploadContext = default;
            void* mappedData = null;
            if (description.CpuAccess == CpuAccessMode.Write)
            {
                mappedData = pMappedData;
            }
            else
            {
                uploadContext = _device.Allocate(allocatedSize);
                mappedData = uploadContext.UploadBuffer.pMappedData;
            }

            for (uint i = 0; i < numSubResources; ++i)
            {
                if (_rowSizesInBytes[i] > unchecked((ulong)-1))
                    continue;

                D3D12_MEMCPY_DEST DestData = new();
                DestData.pData = (void*)((ulong)mappedData + _footprints[i].Offset);
                DestData.RowPitch = _footprints[i].Footprint.RowPitch;
                DestData.SlicePitch = _footprints[i].Footprint.RowPitch * _numRows[i];
                MemcpySubresource(&DestData, &data[i], (nuint)_rowSizesInBytes[i], _numRows[i], _footprints[i].Footprint.Depth);

                if (uploadContext.IsValid)
                {
                    D3D12_TEXTURE_COPY_LOCATION Dst = new(_handle, i);
                    D3D12_TEXTURE_COPY_LOCATION Src = new(uploadContext.UploadBuffer!.Handle, _footprints[i]);
                    uploadContext.CommandList.Get()->CopyTextureRegion(
                        &Dst,
                        0,
                        0,
                        0,
                        &Src,
                        null
                    );
                }
            }

            if (uploadContext.IsValid)
            {
                _device.Submit(in uploadContext);
            }
        }

        State = description.InitialLayout;
    }

    public D3D12Texture(D3D12GraphicsDevice device, ID3D12Resource* existingTexture, in TextureDescription description)
        : base(description)
    {
        _device = device;
        DxgiFormat = (DXGI_FORMAT)description.Format.ToDxgiFormat();
        _handle = existingTexture;
        State = description.InitialLayout;

        if (!string.IsNullOrEmpty(description.Label))
        {
            OnLabelChanged(description.Label!);
        }
    }

    /// <inheritdoc />
    public override GraphicsDevice Device => _device;

    public DXGI_FORMAT DxgiFormat { get; }
    public ID3D12Resource* Handle => _handle;
    public ResourceStates State { get; set; }
    public ResourceStates TransitioningState { get; set; } = (ResourceStates)uint.MaxValue;
    public ulong AllocatedSize { get; }

    /// <summary>
    /// Finalizes an instance of the <see cref="D3D12Texture" /> class.
    /// </summary>
    ~D3D12Texture() => Dispose(disposing: false);

    /// <inheitdoc />
    protected internal override void Destroy()
    {
        foreach (DescriptorIndex index in _RTVs.Values)
        {
            _device.RenderTargetViewHeap.ReleaseDescriptors(index);
        }
        _RTVs.Clear();

        foreach (DescriptorIndex index in _DSVs.Values)
        {
            _device.DepthStencilViewHeap.ReleaseDescriptors(index);
        }
        _DSVs.Clear();

        if (_sharedHandle.Value != null)
        {
            _ = CloseHandle(_sharedHandle);
        }

        _allocation.Dispose();
        _handle.Dispose();
        Free(_footprints);
        Free(_rowSizesInBytes);
        Free(_numRows);
    }

    /// <inheritdoc />
    protected override void OnLabelChanged(string newLabel)
    {
        fixed (char* pName = newLabel)
        {
            _ = _handle.Get()->SetName((ushort*)pName);
        }
    }

    public D3D12_CPU_DESCRIPTOR_HANDLE GetRTV(int mipSlice, int arraySlice, DXGI_FORMAT format = DXGI_FORMAT_UNKNOWN)
    {
        int hash = HashCode.Combine(mipSlice, arraySlice, format);

        if (!_RTVs.TryGetValue(hash, out DescriptorIndex descriptorIndex))
        {
            D3D12_RESOURCE_DESC desc = _handle.Get()->GetDesc();

            D3D12_RENDER_TARGET_VIEW_DESC viewDesc = default;
            viewDesc.Format = format;
            viewDesc.ViewDimension = D3D12_RTV_DIMENSION_TEXTURE2D;

            descriptorIndex = _device.RenderTargetViewHeap.AllocateDescriptors(1u);
            _RTVs.Add(hash, descriptorIndex);

            D3D12_CPU_DESCRIPTOR_HANDLE cpuHandle = _device.RenderTargetViewHeap.GetCpuHandle(descriptorIndex);
            _device.Handle->CreateRenderTargetView(_handle.Get(), &viewDesc, cpuHandle);
        }

        return _device.RenderTargetViewHeap.GetCpuHandle(descriptorIndex);
    }

    public D3D12_CPU_DESCRIPTOR_HANDLE GetDSV(int mipSlice, int arraySlice, DXGI_FORMAT format = DXGI_FORMAT_UNKNOWN)
    {
        int hash = HashCode.Combine(mipSlice, arraySlice, format);

        if (!_DSVs.TryGetValue(hash, out DescriptorIndex descriptorIndex))
        {
            D3D12_RESOURCE_DESC desc = _handle.Get()->GetDesc();

            D3D12_DEPTH_STENCIL_VIEW_DESC viewDesc = default;
            viewDesc.Format = format;
            viewDesc.ViewDimension = D3D12_DSV_DIMENSION_TEXTURE2D;

            descriptorIndex = _device.DepthStencilViewHeap.AllocateDescriptors(1u);
            _DSVs.Add(hash, descriptorIndex);

            D3D12_CPU_DESCRIPTOR_HANDLE cpuHandle = _device.DepthStencilViewHeap.GetCpuHandle(descriptorIndex);
            _device.Handle->CreateDepthStencilView(_handle.Get(), &viewDesc, cpuHandle);
        }

        return _device.DepthStencilViewHeap.GetCpuHandle(descriptorIndex);
    }
}
